modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java

```diff
@@ -322,6 +324,33 @@ private void _validateDDMFormFieldPropertyValue(
 		}
 	}
 
+	private void _validateDDMFormFieldReferences(
+			List<DDMFormField> ddmFormFields)
+		throws DDMFormValidationException {
+
+		if (ddmFormFields.isEmpty()) {
+			return;
+		}
+
+		Set<String> ddmFormFieldReferences = new HashSet<>();
+		Set<String> duplicateDDMFormFieldReferences = new HashSet<>();
+
+		for (DDMFormField ddmFormField : ddmFormFields) {
+			if (!ddmFormFieldReferences.add(ddmFormField.getFieldReference())) {
+				duplicateDDMFormFieldReferences.add(
+					ddmFormField.getFieldReference());
+			}
+
+			_validateDDMFormFieldReferences(
+				ddmFormField.getNestedDDMFormFields());
```

Comment:
@devpedrotavares since you are not using the same `duplicateDDMFormFieldReferences` when validating the `nestedDDMFormFields`, the following scenario will not work:

[Gravação de tela de 09-12-2022 09:21:58.webm](https://user-images.githubusercontent.com/34186632/206701747-8e04dbdc-09f8-4c1e-b5b4-88053af5c276.webm)

Reply:
I would suggest something like:

```
_validateDDMFormFieldReferences(
	ddmFormFields, new HashSet<>(), new HashSet<>());

...

private void _validateDDMFormFieldReferences(
		List<DDMFormField> ddmFormFields,
		Set<String> ddmFormFieldReferences,
		Set<String> duplicateDDMFormFieldReferences)
	throws DDMFormValidationException {

	for (DDMFormField ddmFormField : ddmFormFields) {
		if (!ddmFormFieldReferences.add(ddmFormField.getFieldReference())) {
			duplicateDDMFormFieldReferences.add(
				ddmFormField.getFieldReference());
		}

		_validateDDMFormFieldReferences(
			ddmFormField.getNestedDDMFormFields(), ddmFormFieldReferences,
			duplicateDDMFormFieldReferences);
	}

	if (SetUtil.isNotEmpty(duplicateDDMFormFieldReferences)) {
		throw new MustNotDuplicateFieldReference(
			duplicateDDMFormFieldReferences);
	}
}
```

Could you take a look at this?

Reply:
That's right! I did something similar. But I created a helper method, so that all the duplicated references caught will be added to the set.